sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1248 @@
+                        intercept_grads, layer_units, incremental):
+        rng = check_random_state(self.random_state)
+
+        if not incremental or not hasattr(self, '_optimizer'):
+            params = self.coefs_ + self.intercepts_
+            optimizer = stochastic_optimizers[self.algorithm]
+            self._optimizer = optimizer(params, self.learning_rate_init,
+                                        lr_schedule=self.learning_rate,
+                                        momentum=self.momentum,
+                                        nesterov=self.nesterovs_momentum,
+                                        power_t=self.power_t,
+                                        beta_1=self.beta_1, beta_2=self.beta_2,
+                                        epsilon=self.epsilon)
+
+        # early_stopping in partial_fit doesn't make sense
+        early_stopping = self.early_stopping and not incremental
+        if early_stopping:
+            X, X_val, y, y_val = train_test_split(
+                X, y, random_state=self.random_state,
+                test_size=self.validation_fraction)
+            if isinstance(self, ClassifierMixin):
+                y_val = self.label_binarizer_.inverse_transform(y_val)
+        else:
+            X_val = None
+            y_val = None
+
+        n_samples = X.shape[0]
+        batch_size = np.clip(self.batch_size, 1, n_samples)
+
+        try:
+            for it in range(self.max_iter):
+                X, y = shuffle(X, y, random_state=rng)
+                accumulated_loss = 0.0
+                for batch_slice in gen_batches(n_samples, batch_size):
+                    activations[0] = X[batch_slice]
+                    batch_loss, coef_grads, intercept_grads = self._backprop(
+                        X[batch_slice], y[batch_slice],
+                        activations, deltas, coef_grads,
+                        intercept_grads)
+                    accumulated_loss += batch_loss * (batch_slice.stop -
+                                                      batch_slice.start)
+
+                    # update weights
+                    grads = coef_grads + intercept_grads
+                    self._optimizer.update_params(grads)
+
+                self.n_iter_ += 1
+                self.loss_ = accumulated_loss / X.shape[0]
+
+                self.t_ += n_samples
+                self.loss_curve_.append(self.loss_)
+                if self.verbose:
+                    print("Iteration %d, loss = %.8f" % (self.n_iter_,
+                                                         self.loss_))
+
+                # update no_improvement_count based on training loss or
+                # validation score according to early_stopping
+                self._update_no_improvement_count(early_stopping, X_val, y_val)
+
+                # for learning rate that needs to be updated at iteration end
+                self._optimizer.iteration_ends(self.t_)
+
+                if self._no_improvement_count > 2:
```

Comment:
I have seen some patience stuff in the deep learning tutorial but it is update based rather than epoch based - that works pretty well. I am ok with this heuristic but we should be explicit in stating it is a heuristic, and might consider exposing it as yet another hyperparameter (though the tolerance kind of interacts with this, so it is exposed indirectly)